File: LAB_PARCIAL/REV4/CLIENTE_MICRO_CHUNK-PYTHON/tcp_client_50.py
import socket
import threading

class TCPClient50:
    DEFAULT_SERVER_PORT = 4444

    def __init__(self, ip, port=DEFAULT_SERVER_PORT, callback=None):
        self.server_ip = ip
        self.server_port = port
        self.listeners = []
        self.callback = callback
        self.running = False
        self.connected = False
        self.socket = None
        self.lock = threading.Lock()

    def add_listener(self, listener):
        if listener not in self.listeners:
            self.listeners.append(listener)

    def remove_listener(self, listener):
        if listener in self.listeners:
            self.listeners.remove(listener)

    def is_connected(self):
        return self.connected

    def send_message(self, message):
        if self.socket and self.connected:
            try:
                self.socket.sendall((message + "\n").encode('utf-8'))
                return True
            except:
                self.connected = False
                self.notify_error("Conexión perdida al enviar.")
                return False
        else:
            self.notify_error("No hay conexión activa con el servidor.")
            return False

    def stop_client(self):
        self.running = False
        self.connected = False
        if self.socket:
            try:
                self.socket.shutdown(socket.SHUT_RDWR)
                self.socket.close()
            except: 
                pass
        self.socket = None

    def run(self):
        self.running = True
        try:
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.socket.settimeout(5.0)
            self.socket.connect((self.server_ip, self.server_port))
            self.socket.settimeout(None)
            self.connected = True
            self.notify_connected()

            with self.socket.makefile('r', encoding='utf-8') as reader:
                while self.running:
                    line = reader.readline()
                    if not line: 
                        break
                    clean_line = line.strip()
                    if clean_line:
                        self.notify_message(clean_line)
        except Exception as e:
            if self.running: 
                self.notify_error(f"Fallo de red: {e}")
        finally:
            self.stop_client()
            self.notify_disconnected()

    def notify_message(self, msg):
        for l in list(self.listeners):
            if hasattr(l, 'message_received'): 
                l.message_received(msg)

    def notify_connected(self):
        for l in list(self.listeners):
            if hasattr(l, 'on_connected'): 
                l.on_connected()

    def notify_disconnected(self):
        for l in list(self.listeners):
            if hasattr(l, 'on_disconnected'): 
                l.on_disconnected()

    def notify_error(self, msg):
        for l in list(self.listeners):
            if hasattr(l, 'on_error'): 
                l.on_error(msg)

File: LAB_PARCIAL/REV4/CLIENTE_MICRO_CHUNK-JAVA/redesOk/TCPClient50.java
package redesOk;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.net.InetAddress;
import java.net.Socket;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;


//Cliente TCP persistente, mantiene una conexion abierta con el servidor
//y permite que varias ventanas escuchen respuestas sin crear nuevas conexiones
public class TCPClient50 {
    public static final int DEFAULT_SERVER_PORT = 4444;

    private final String serverIp;
    private final int serverPort;
    private final List<OnMessageReceived> listeners = new CopyOnWriteArrayList<>();

    private volatile boolean running = false;
    private volatile boolean connected = false;

    private PrintWriter out;
    private BufferedReader in;
    private Socket socket;

    public TCPClient50(String ip, OnMessageReceived listener) {
        this(ip, DEFAULT_SERVER_PORT, listener);
    }

    public TCPClient50(String ip, int port, OnMessageReceived listener) {
        this.serverIp = ip;
        this.serverPort = port;
        if (listener != null) {
            listeners.add(listener);
        }
    }

    public void addListener(OnMessageReceived listener) {
        if (listener != null) {
            listeners.add(listener);
        }
    }

    public void removeListener(OnMessageReceived listener) {
        listeners.remove(listener);
    }

    public boolean isConnected() {
        return connected && out != null && !out.checkError();
    }

    public void sendMessage(String message) {
        if (out != null && !out.checkError()) {
            out.println(message);
            out.flush();
        } else {
            notifyError("No hay conexion activa con el servidor.");
        }
    }

    public void stopClient() {
        running = false;
        connected = false;
        try {
            if (socket != null) {
                socket.close();
            }
        } catch (Exception ignored) {
        }
    }

    public void run() {
        running = true;

        try {
            InetAddress serverAddr = InetAddress.getByName(serverIp);
            socket = new Socket(serverAddr, serverPort);

            out = new PrintWriter(
                    new BufferedWriter(new OutputStreamWriter(socket.getOutputStream())),
                    true
            );
            in = new BufferedReader(new InputStreamReader(socket.getInputStream()));

            connected = true;
            notifyConnected();

            String serverMessage;
            while (running && (serverMessage = in.readLine()) != null) {
                notifyMessage(serverMessage);
            }
        } catch (Exception e) {
            notifyError("Error conectando a " + serverIp + ":" + serverPort + " -> " + e.getMessage());
        } finally {
            running = false;
            connected = false;
            try {
                if (socket != null) {
                    socket.close();
                }
            } catch (Exception ignored) {
            }
            notifyDisconnected();
        }
    }

    private void notifyMessage(String message) {
        for (OnMessageReceived listener : listeners) {
            listener.messageReceived(message);
        }
    }

    private void notifyConnected() {
        for (OnMessageReceived listener : listeners) {
            listener.connected();
        }
    }

    private void notifyDisconnected() {
        for (OnMessageReceived listener : listeners) {
            listener.disconnected();
        }
    }

    private void notifyError(String message) {
        for (OnMessageReceived listener : listeners) {
            listener.error(message);
        }
    }

    public interface OnMessageReceived {
        void messageReceived(String message);

        default void connected() {
        }

        default void disconnected() {
        }

        default void error(String message) {
        }
    }
}
